Source/Scene/Cesium3DTilesetTraversal.js

```diff
@@ -1,856 +1,643 @@
+        // If skipLevelOfDetail is off then a tile does not refine until all children are loaded.
+        // This is the traditional replacement refinement approach and is called the base traversal.
+        // Tiles that have a greater screen space error than the base screen space error are part of the base traversal,
+        // all other tiles are part of the skip traversal. The skip traversal allows for skipping levels of the tree
+        // and rendering children and parent tiles simultaneously.
+        var stack = traversal.stack;
+        stack.push(root);
 
-    function computeChildrenVisibility(tile, frameState) {
-        var flag = Cesium3DTileChildrenVisibility.NONE;
-        var children = tile.children;
-        var childrenLength = children.length;
-        var visibilityPlaneMask = tile._visibilityPlaneMask;
-        for (var k = 0; k < childrenLength; ++k) {
-            var child = children[k];
+        while (stack.length > 0) {
+            traversal.stackMaximumLength = Math.max(traversal.stackMaximumLength, stack.length);
 
-            var visibilityMask = child.visibility(frameState, visibilityPlaneMask);
+            var tile = stack.pop();
+            var baseTraversal = inBaseTraversal(tileset, tile, baseScreenSpaceError);
+            var add = tile.refine === Cesium3DTileRefine.ADD;
+            var replace = tile.refine === Cesium3DTileRefine.REPLACE;
+            var children = tile.children;
+            var childrenLength = children.length;
+            var parent = tile.parent;
+            var parentRefines = !defined(parent) || parent._refines;
+            var traverse = (childrenLength > 0) && (tile._screenSpaceError > maximumScreenSpaceError);
+            var refines = false;
+
+            if (tile.hasTilesetContent && tile.contentExpired) {
+                // Don't traverse expired subtree because it will be destroyed
+                traverse = false;
+            }
 
-            if (isVisible(visibilityMask)) {
-                flag |= Cesium3DTileChildrenVisibility.VISIBLE;
+            if (traverse) {
+                refines = updateAndPushChildren(tileset, tile, stack, frameState) && parentRefines;
             }
 
-            if (!child.insideViewerRequestVolume(frameState)) {
-                if (isVisible(visibilityMask)) {
-                    flag |= Cesium3DTileChildrenVisibility.VISIBLE_NOT_IN_REQUEST_VOLUME;
-                }
-                visibilityMask = CullingVolume.MASK_OUTSIDE;
-            } else {
-                flag |= Cesium3DTileChildrenVisibility.IN_REQUEST_VOLUME;
-                if (isVisible(visibilityMask)) {
-                    flag |= Cesium3DTileChildrenVisibility.VISIBLE_IN_REQUEST_VOLUME;
+            if (hasEmptyContent(tile)) {
+                // Add empty tile just to show its debug bounding volume
+                addEmptyTile(tileset, tile, frameState);
+                loadTile(tileset, tile, frameState);
```

Comment:
Why are we loading an empty tile? Won't it just return from the `loadTile` function?

Reply:
`hasEmptyContent` is a catch-all for anything that has `Empty3DTileContent` or `Tileset3DTileContent`. The load is needed for the `Tileset3DTileContent` case to load the external tileset. I left a comment to clarify.